src/main/java/org/broadinstitute/hellbender/tools/dataflow/transforms/bqsr/BaseRecalibratorFn.java

```diff
@@ -81,167 +59,118 @@
-        this.BRAC = toolArgs;
+    // saves to BRAC.BRAC.RAC.RECAL_TABLE_FILE.getName()
+    public static void SaveTextualReport(File output, SAMFileHeader header, RecalibrationTables rt, BaseRecalibrationArgumentCollection BRAC) throws IOException {
+        QuantizationInfo qi = new QuantizationInfo(rt, BRAC.RAC.QUANTIZING_LEVELS);
+        BaseRecalibratorFn worker = new BaseRecalibratorFn(header, BRAC);
+        worker.onTraversalStart();
+        worker.saveReport(output, rt, qi, worker.covariates);
     }
 
-    /**
-     * Throws if knownSites is missing.
-     */
-    public void checkClientArguments() {
-        if (null == BRAC.RAC.knownSites || BRAC.RAC.knownSites.size() == 0) {
-            throw new UserException.CommandLineException(NO_DBSNP_EXCEPTION);
-        }
+    @Override
+    public void startBundle(Context c) throws Exception {
+        super.startBundle(c);
+        // can't set things up here because we don't have our side input yet.
+        // So instead we're doing it in processElement.
+        nReadsProcessed = 0;
+        // resetting firstInBundle in case we have multiple bundles via the same instance.
+        firstInBundle = true;
     }
 
-    public void onTraversalStart(File refFile) {
-        accumulator = 0L;
-
-        baq = new BAQ(BRAC.BAQGOP); // setup the BAQ object with the provided gap open penalty
-
-        if (BRAC.RAC.FORCE_PLATFORM != null)
-            BRAC.RAC.DEFAULT_PLATFORM = BRAC.RAC.FORCE_PLATFORM;
+    @Override
+    public void processElement(ProcessContext c) throws Exception {
 
-        covariates = new StandardCovariateList(BRAC.RAC, getHeaderForReads());
-
-        initializeRecalibrationEngine();
-        minimumQToUse = BRAC.PRESERVE_QSCORES_LESS_THAN;
-
-        if (null!=refFile) {
-            referenceDataSource = new ReferenceDataSource(refFile);
+        // can't set things up in startBundle because it doesn't have our side input yet.
+        // So instead we're doing it here in processElement.
+        if (firstInBundle) {
+            init(c);
+            onTraversalStart();
+            firstInBundle = false;
         }
+        GATKRead r = c.element().getKey();
+        ReadContextData rc = c.element().getValue();
+        Iterable<Variant> variants = rc.getOverlappingVariants();
+        final ReferenceBases refBases = rc.getOverlappingReferenceBases();
+        ReferenceDataSource refDS = new ReferenceDataSourceFromReferenceBases(refBases, referenceSequenceDictionary);
+        ReferenceContext ref = new ReferenceContext(refDS, refBases.getInterval());
+
+        apply(r, ref, refDS, variants);
+    }
+
+    @Override
+    public void finishBundle(Context c) throws Exception {
+        bunny.stepEnd("finishBundle");
+        if (null!=recalibrationEngine) {
+            final RecalibrationTables recalibrationTables = recalibrationEngine.getRecalibrationTables();
+            if (null==recalibrationTables) {
+                bunny.stepEnd("null output(!)");
```

Comment:
This is a "should never happen" line but at the same time it would have no deleterious effect as the rest of the code correctly interprets a null there as an empty table. For this cases I generally log something but don't force a failure. But this is a matter of policy, and we can follow whichever policy you think is best.